File: Java/Maximum-Product-Subarray.java
public class Solution {
    /**
     * @param nums: an array of integers
     * @return: an integer
     */
    public int maxProduct(int[] nums) {
        // write your code here
        int cur = nums[0], maxv = nums[0], minv = nums[0];
        int x, y, z;
        for (int i = 1; i < nums.length; i ++) {
            x = nums[i] * maxv;
            y = nums[i] * minv;
            z = nums[i];
            minv = Math.min(x, Math.min(y, z));
            maxv = Math.max(x, Math.max(y, z));
            if (maxv > cur) cur = maxv;
        }
        return cur;
    }
}

File: Cpp/Maximum-Product-Subarray.cpp
#include <vector>

class Solution {
public:
    /**
     * @param nums: a vector of integers
     * @return: an integer
     */
    int maxProduct(vector<int>& nums) {
        // write your code here
        int minp = nums[0], maxp = nums[0], cnt = nums[0];
        int x, y, z;
        for (int i = 1; i < nums.size(); i ++) {
            x = nums[i] * minp;
            y = nums[i] * maxp;
            z = nums[i];
            minp = min(min(x, y), z);
            maxp = max(max(x, y), z);
            if (maxp > cnt) cnt = maxp;
        }
        return cnt;
    }
};
